Add stackallocated Span<char> in Format method

YuriiBabenko opened the pull request:
I added using stack allocated Spans of char instead of allocating new strings or using StringBuilder. It allows reduce heap allocations + speeds it a bit. I ran a benchmark to compare how library and my version formats phone number in `PhoneNumberFormat.E164` format. Screenshot from benchmark below:

![image](https://user-images.githubusercontent.com/26046889/231803782-66909794-4d9a-4acd-ac7e-614f0285cc51.png)

There is still space to optimize it a bit (e.g. `FormatNsn` method). So, if you are interested in it, please, left a comment in this PR and I'll continue working on it.

yurii-babenko96-ideals committed:
Add stackallocated Span<char> in Format method

twcclegg commented on csharp/PhoneNumbers/PhoneNumberUtil.cs:
```diff
@@ -1200,7 +1201,107 @@ private bool HasValidCountryCallingCode(int countryCallingCode)
+        /// <param name="numberFormat">The format the phone number should be formatted into.</param>
+        /// <returns>The formatted phone number.</returns>
+        public string Format(PhoneNumber number, PhoneNumberFormat numberFormat)
+        {
```
As to not be breaking `public void Format(PhoneNumber number, PhoneNumberFormat numberFormat, StringBuilder formattedNumber)` should still exists.

YuriiBabenko replied:
yeah, you are right, i forgot to move it outside `#else` statement. i'll fix it in the next commit

YuriiBabenko replied:
should be fine now

twcclegg commented:
Can you update the test project to retain coverage of the non-span version, ie https://github.com/twcclegg/libphonenumber-csharp/pull/167

yurii-babenko96-ideals committed:
Minor fixes

yurii-babenko96-ideals committed:
Increase test coverage

yurii-babenko96-ideals committed:
Another way to increase test coverage

YuriiBabenko commented:
@twcclegg I added tests to use `Format(PhoneNumber number, PhoneNumberFormat numberFormat, StringBuilder formattedNumber)` directly and added `netframework4.8` as a test target framework. i guess, more tests is not less tests, so it should be fine, but i'll remove them, if it's not ok.

also, as I understand,  you are interested in this optimization, but could you confirm that, please? :)

yurii-babenko96-ideals committed:
Add netframework as target to Extension.Tests project

YuriiBabenko commented:
@twcclegg also i'd like to add `net6.0` to target frameworks list, because it is LTS version.

what do you think about it?

yurii-babenko96-ideals committed:
Move different approaches to different parts of partial class

yurii-babenko96-ideals committed:
Add net6.0 as target framework

YuriiBabenko commented:
also i rearranged code in `PhoneNumberUtil.cs` file, because it had everything mixed, it was hard to read. but i see that now this file contain 4k lines of changes and it does not look like a good idea anymore, so i'll revert it later, unless you want to keep it that way

yurii-babenko96-ideals committed:
Optimize 3 more methods

yurii-babenko96-ideals committed:
revert code rearranging in PhoneNumberUtil.cs file

yurii-babenko96-ideals committed:
Add nullable for net6.0 target framework

twcclegg commented on csharp/PhoneNumbers/PhoneNumberUtil.netcore.cs:
```diff
@@ -0,0 +1,792 @@
+﻿using System;
```
nit: `.net` rather than `.netcore`

YuriiBabenko replied:
done

twcclegg commented on csharp/PhoneNumbers.Extensions/PhoneNumberConverter.cs:
```diff
@@ -15,4 +16,20 @@ public class PhoneNumberConverter : JsonConverter<PhoneNumbers.PhoneNumber>
             => writer.WriteStringValue(Util.Format(value, PhoneNumberFormat.E164));
     }
+
+    public class PhoneNumberConverter_Old : JsonConverter<PhoneNumbers.PhoneNumber>
```
What's the benefit of this?
It uses the StringBuilder version, but don't know that that's compelling.

YuriiBabenko replied:
it was before i added `netframework4.8` as a target framework. i'll clean it up

YuriiBabenko replied:
done

twcclegg commented on csharp/PhoneNumbers.Extensions.Test/TestConversion.cs:
```diff
@@ -33,13 +60,84 @@ public void TestSerialization(string expected, string input, string region)
             var json = $$"""{"PhoneNumber": "{{value}}"}""";
+#else
+            var json = $@"{{""PhoneNumber"": ""{value}""}}";
+#endif
```
Is there value to this split-track?

YuriiBabenko replied:
i was not sure, so i kept it. there should have been some reason to use raw strings, but, imho, no. i'll remove it

YuriiBabenko replied:
done

twcclegg commented on csharp/PhoneNumbers/PhoneNumberUtil.netcore.cs:
```diff
@@ -0,0 +1,792 @@
+{
+    public partial class PhoneNumberUtil
+    {
+#if NET5_0_OR_GREATER || NETSTANDARD2_1_OR_GREATER
```
Is there value to have this for just the class contents rather than the entire file?

YuriiBabenko replied:
nope, fixed

YuriiBabenko commented:
@twcclegg no idea what is the problem with code coverage. `netframewok4.8` tests call this methods. that's screenshot from my rider with test coverage for all target frameworks
![image](https://user-images.githubusercontent.com/26046889/232223582-71f59112-3c24-4838-8d0d-3782024c4d41.png)

twcclegg reviewed (commented):
A couple of small things, other than that should be ready to merge.

yurii-babenko96-ideals committed:
PR fixes

YuriiBabenko commented:
@twcclegg I made requested updates. actual code coverage below:
![image](https://user-images.githubusercontent.com/26046889/232691229-62542720-8b31-4a96-885b-ebf7f12cc78e.png)

who should resolve your comments to PR, you or I?

yurii-babenko96-ideals committed:
remove duplicate tests
